community/lucene-index/src/main/java/org/neo4j/kernel/api/impl/index/storage/paged/FSALockFactory.java

```diff
@@ -0,0 +1,232 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.api.impl.index.storage.paged;
+
+import org.apache.lucene.store.AlreadyClosedException;
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.Lock;
+import org.apache.lucene.store.LockFactory;
+import org.apache.lucene.store.LockObtainFailedException;
+import org.apache.lucene.store.NativeFSLockFactory;
+import org.apache.lucene.util.IOUtils;
+
+import java.io.IOException;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
+import org.neo4j.io.fs.FileSystemAbstraction;
+import org.neo4j.io.fs.OpenMode;
+import org.neo4j.io.fs.StoreChannel;
+
+/**
+ * Clone of {@link NativeFSLockFactory} for {@link FileSystemAbstraction}.
+ */
+public class FSALockFactory extends LockFactory
+{
+    // Lucene implements concurrency across multiple Directory instances
```

Comment:
Agree with this; doing the only-in-process locking is what `SingleInstanceLockFactory` does. However, it felt a bit scary to make more than one major change at a time with Lucene - ie. I opted to keep the locking the same as we currently have (file locking via NativeFSLockFactory), so that if we hit issues with the changes this patch introduces there are fewer moving parts.

All to say: I agree, but think we should wait to make the change to in-process locking separately from this PR.